### PR TITLE
[FAB-17906] Implement snapshotRequestBookKeeper

### PR DESCRIPTION
**Pull request by wenjianqiao**
Signed-off-by: Wenjian Qiao <wenjianq@gmail.com>

#### Type of change
- New feature

#### Description
Add a bookkeeper db to store snapshot requests and add/delete/list
snapshot requests. It also maintains smallest height for all the
pending snapshot requests.

#### Additional details
This is part of FAB-17906. In a separate PR, the snapshot mgmt APIs will be implemented and call the snapshot request bookkeeper to submit/cancel/list snapshot requests.

#### Related issues
https://jira.hyperledger.org/browse/FAB-17906

### CODE REVIEW

File: core/ledger/kvledger/snapshot_mgmt.go [hunk 1]
```diff
@@ -60,3 +62,130 @@ func (l *kvLedger) ListSnapshots() ([]string, error) {
...
+func (bk *snapshotRequestBookkeeper) smallestRequest() (uint64, error) {
+	itr, err := bk.dbHandle.GetIterator(nil, nil)
+	if err != nil {
+		return 0, err
+	}
+	defer itr.Release()
+
+	smallestRequestHeight := defaultSmallestHeight
+	if itr.Next() {
+		if err := itr.Error(); err != nil {
```
Comment (manish-sethi) [inline core/ledger/kvledger/snapshot_mgmt.go]: could cause miss of an error

File: core/ledger/kvledger/snapshot_mgmt.go [hunk 2]
```diff
@@ -60,3 +62,130 @@ func (l *kvLedger) ListSnapshots() ([]string, error) {
...
+	var err error
+	if bk.smallestRequestHeight, err = bk.smallestRequest(); err != nil {
+		return nil, err
+	}
+
+	return bk, nil
+}
+
+// add adds the given height to the bookkeeper db and returns an error if the height already exists
+func (bk *snapshotRequestBookkeeper) add(height uint64) error {
```
Comment (manish-sethi) [inline core/ledger/kvledger/snapshot_mgmt.go]: Nit picking: Idiomatically, you would use one letter name for the function receiver. You may like to change `bk` to `b` or `k`

File: core/ledger/kvledger/snapshot_mgmt_test.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	gbHash := protoutil.BlockHeaderHash(gb.Header)
+	l, err := provider.Create(gb)
+	require.NoError(t, err)
+	defer l.Close()
+	kvl := l.(*kvLedger)
+
+	bcInfo, _ := l.GetBlockchainInfo()
+	require.Equal(t, &common.BlockchainInfo{
+		Height: 1, CurrentBlockHash: gbHash, PreviousBlockHash: nil,
+	}, bcInfo)
```
Comment (manish-sethi) [inline core/ledger/kvledger/snapshot_mgmt_test.go]: Though this code is not related and you could directly create an instance. However, since you have it here, it will be good to add a case for error propagation during ledger opening.

File: core/ledger/kvledger/snapshot_mgmt_test.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+	require.NoError(t, kvl2.snapshotRequestBookkeeper.delete(50))
+	require.Equal(t, defaultSmallestHeight, kvl2.snapshotRequestBookkeeper.smallestRequestHeight)
+
+	requestHeights, err = kvl2.snapshotRequestBookkeeper.list()
+	require.NoError(t, err)
+	require.ElementsMatch(t, requestHeights, []uint64{})
+}
+
+func TestSnapshotRequestBookKeeper_ErrorPaths(t *testing.T) {
```
Comment (manish-sethi) [inline core/ledger/kvledger/snapshot_mgmt_test.go]: Nit: `_` in function name is not idiomatic

**Comment by wenjianqiao** [comment]
@manish-sethi Thank you. I have addressed your comments and also simplified unit tests by directly creating a snapshotRequestBookkeeper. The unit test to kvledger change will be covered when snapshot mgmt APIs are implemented.

**Review by manish-sethi — APPROVED** [review]
